// @electron-forge/maker-msix/dist/MakerMSIX.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MakerMSIX = void 0;
const node_os_1 = __importDefault(require("node:os"));
const node_path_1 = __importDefault(require("node:path"));
const core_utils_1 = require("@electron-forge/core-utils");
const maker_base_1 = require("@electron-forge/maker-base");
const electron_windows_msix_1 = require("electron-windows-msix");
const fs_extra_1 = __importDefault(require("fs-extra"));
const arch_1 = require("./util/arch");
/**
 * Creates an MSIX package for your Electron app.
 * @experimental
 */
class MakerMSIX extends maker_base_1.MakerBase {
    constructor() {
        super(...arguments);
        this.name = 'msix';
        this.defaultPlatforms = ['win32'];
    }
    isSupportedOnCurrentPlatform() {
        return process.platform === 'win32';
    }
    async make({ dir, makeDir, targetArch, packageJSON, appName, }) {
        const { manifestVariables, ...packageOptions } = this.config;
        // Do all the scratch work in a temporary folder
        const tmpFolder = await fs_extra_1.default.mkdtemp(node_path_1.default.resolve(node_os_1.default.tmpdir(), 'msix-maker-'));
        try {
            const result = await (0, electron_windows_msix_1.packageMSIX)({
                ...packageOptions,
                manifestVariables: {
                    packageDescription: packageJSON.description,
                    appExecutable: `${appName}.exe`,
                    packageVersion: packageJSON.version,
                    publisher: (0, core_utils_1.getNameFromAuthor)(packageJSON.author),
                    packageIdentity: appName,
                    targetArch: (0, arch_1.toMsixArch)(targetArch),
                    ...manifestVariables,
                },
                appDir: dir,
                outputDir: tmpFolder,
            });
            const outputPath = node_path_1.default.resolve(makeDir, 'msix', targetArch, `${appName}.msix`);
            await fs_extra_1.default.mkdirp(node_path_1.default.dirname(outputPath));
            await fs_extra_1.default.move(result.msixPackage, outputPath);
            return [outputPath];
        }
        finally {
            fs_extra_1.default.remove(tmpFolder);
        }
    }
}
exports.default = MakerMSIX;
exports.MakerMSIX = MakerMSIX;
